### DISCUSSION TITLE
Welcome to DINEOF Discussions!

### DISCUSSION THREAD

**Discussion by aida-alvera** [aida-alvera/DINEOF#12] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members using DINEOF and also as Q&A forum. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Tom933** [comment]
hello, When I use DINEOF, the program is directly interrupted when reading gappy data and mask. What is the reason? 
It's the step.
You entered filenames sst.avhrr
                       sst.avhrr.mask

**Reply by ctroupin** [reply]
Hello, could you provide more details about the issue, for instance copy and paste the error message you get?
Could you properly compile the code? 

Also maybe it is better to open an issue.
Thanks.

**Reply by Tom933** [reply]
I am using the compiled file directly.Under Windows 10.I am not sure where to find the error message. This problem is the first time I have encountered it, and it has been used before.

**Reply by ctroupin** [reply]
Could you run the `Small example` provided with the code (in `DINEOF/SmallExample/`)?
```bash
../dineof dineof.init
```
which should output something starting with:
```bash
********************************************************************
 Numerical data read
 You entered the values:
 number of EOF modes you want to compute           5
 maximal size for the Krylov subspace          10

 You asked not to normalise the input matrices

 The right and left EOFs will be written in directory Output/

 ********************************************************************

 You entered filenames alboranSST_small.nc#SST
                       alboranSST_small.nc#mask

 initfilename: dineof.init                            
```

**Reply by Tom933** [reply]
yes,I can run the small example.But When I run a file generated by a large array, the program just aborts.
You entered filenames sst.avhrr
                       sst.avhrr.mask
After this step the program stops running，is it because the file is too big?

**Comment by gkinglisboa** [comment]
Hi,

The following links in FAQ.md no longer work.  Also a search for "metR" did not find anything, and the Ref [1] is not in the FAQ doc.

* See also DINEOF implemented in EOFimputation of metR : [1]
* Then there is an online version for use with standard satellite data: http://www.dineof.net/

greg

**Reply by ctroupin** [reply]
Thanks Greg.
I replaced [1] with https://rdrr.io/github/eliocamp/metR/man/ImputeEOF.html,
I think that's the relevant reference for this package.

**Comment by gkinglisboa** [comment]
Thanks.

That looks right.

greg

**Comment by gkinglisboa** [comment]
In the current edition of dineof:  
The SmallExample folder has an init file set up to process the netcdf file 'alboranSST_small.nc'
The output is written to folder Output.  No problem.

You also have a folder named 'Reference'
In that folder you have output from a different example.  

So the output in the Reference folder does not serve as a proper Reference for the example used in the init file.

**Comment by ctroupin** [comment]
Thanks for reporting that issue, we will correct that and add the good reference.

**Comment by gkinglisboa** [comment]
Hi, 
Your scripts directory has many useful matlab routines. However, our institution no longer has a matlab license and we must now port our matlab code to python.  

Do you have scripts written in python? Or must we make the translation ourselves?

greg

**Comment by ctroupin** [comment]
Hi Greg,

did you try working with Octave (which doesn't require a license).
Otherwise I think we have tools written in Julia, another language that we are now using a lot (also free and open: https://julialang.org/), I add @aida-alvera and @Alexander-Barth to see if they can provide a link to the Julia tools.

Of course we can help if you plan to write something in Python, which should not be too complicated.

Charles

**Reply by gkinglisboa** [reply]
Hi Charles,

Python is currently the language used and taught in our oceanography and atmosphere groups and thus best that we use it.  Julia is something new for us and would require a learning curve. 

greg

**Reply by gkinglisboa** [reply]
Charles, 
Our greatest need is to have python replacements for 
gwrite.m 
gread.m

greg

**Reply by ctroupin** [reply]
I move my previous reply (which was in a new thread) here just to be complete

>  OK, it makes sense.
> Maybe what can be done is to create a set of python functions, based on what exists on Julia, so that they are (more or less) doing the same things. If you start a github repository we can help directly there, of needed.


About `gwrite.m` and `gread.m`: yes I imagined! I think I had once their Python version on my machine, I'll try to recover them now.

**Reply by gkinglisboa** [reply]
Thanks Charles!
Meanwhile I will find out how to set up a github repository.  I have not done that before -- I am a complete novice.

greg

**Reply by ctroupin** [reply]
Well, that's just a suggestion, you can do it without a repository, we will help anyway ;)

**Reply by ctroupin** [reply]
OK, I think you can start from this module: https://github.com/gher-ulg/DivaPythonTools/blob/master/pydiva/GHERFormat.py
thanks to @Alexander-Barth who provided me that a long time ago.

**Reply by gkinglisboa** [reply]
Thanks Charles!  I will get back to you asap.

greg

**Reply by Alexander-Barth** [reply]
Glad that this function is still useful. :-) Do not forget to use the correct `endianess` (oh boy, I guess I should have written endianness) corresponding to these options in gfortran (https://gcc.gnu.org/onlinedocs/gfortran/Runtime-Options.html) . Other compilers have similar options. Did you consider to use NetCDF ? (In any case, I have converted uread to julia too, if anybody is interested)

**Reply by Alexander-Barth** [reply]
@aida-alvera instead of having DINEOF/Scripts/*m should we not better have DINEOF/Scripts/matlab/*m, DINEOF/Scripts/python/*py, DINEOF/Scripts/julia/*jl ... so that we have a place where to put the scripts for the other languages?

**Comment by gkinglisboa** [comment]
Thanks for that! 

Let me also mention that we have a short project that successfully installs ARPACK and DINEOF and runs  in a docker container for Ubuntu 20.04.  As soon as I get a reply from our IT staff on the project, I will send you the github repo he set up.

I had not considered netCDF, but that would be a much more satisfactory way to do things. 

Our application is a bit special (but does not prevent using netCDF).  Let me explain.

Using weekly mean SST (pathfinder 5, 5.1, 5.3 data), 
we extract SST values along a “line” close to the coast of Portugal (the middle of the continental shelf). 

This yields a N x 1 column vector of SST values.  
We do this for a given year to produce a N x 52 matrix -- which is our input “image” for dineof. 
Missing and bad data are given the value required by dineof. 
The matrix is written in the appropriate format for import by dineof to produce a filled image.

We repeat this exercise for a line of longitude about 500 km away in the open ocean. 

The difference between the filled midshelf and open ocean lines is calculated to obtain an upwelling index for the Portuguese coast.  

The above is repeated for each year of SST data and then analyzed to obtain an upwelling climatology etc.

Please send me any comments you may have on our project.

Thanks,
greg

**Reply by ctroupin** [reply]
Thanks for sharing, that's interesting.

Maybe a quick question: why do you extract only "one" point? You could certainly use DINEOF on 3D fields (time x lon x lat), then exctract the SST in the middle of the continental shelf and 500 km offshore, to finally estimate the upwelling index.
(tell me if what I wrote is not clear ;)

**Comment by gkinglisboa** [comment]
Charles,

Good point. Perhaps I am stuck in the age when disk space and memory was
much more limited.  I have to rethink the problem.  For the moment we are
keen just to get back on the air, so that means first just repeating what
we did before.  But after that I see advantages to what you suggested.
Some exploration is required.

I am sure you have lots of projects already, but if you are interested in
working with us, that would be welcome.

greg

**Comment by gkinglisboa** [comment]
Charles,

Just to give you an idea, attached are two sst maps (7day composites), one
in winter (Jan, week 2) and one in summer (June, week 26).  The open ocean
line is at 15W and the midshelf line is the curvy line along the coast.

As you can see, there is a lot of missing data in winter.

So your suggestion is to apply dineof to all 52 images (if doing only one
year) or if for 30 years, about 1500 - 2000 images. These days I suppose
that would not be such a big calculation. And then to extract the open
ocean line and the mishelf line from the entire set of images all at once
using only two lines of code.

Do I have that right?

greg

**Reply by ctroupin** [reply]
Hello Greg,

I didn't see the attached images, but anyway, I think it makes more sense to use a longer time series of images, not limited to the coastal band but large enough to properly capture the main features. I would say that the extension of the domain depends on the machine you have: 1500 times steps is already something to deal with, maybe an horizontal grid 100 X 100 grid points is already too much.

**Comment by gkinglisboa** [comment]
Charles 

Here are our github pages for the docker-dineof installations. 
Note -- we may change ownership of these pages in future.

 https://github.com/hugominas/dockerdineofproject <https://github.com/hugominas/dockerdineofproject> 

the docker page https://hub.docker.com/r/hugommineiror/fortrandineof <https://hub.docker.com/r/hugommineiror/fortrandineof> 

greg

**Comment by gkinglisboa** [comment]
Hi Charles,

Thanks -- useful food for thought!

greg

**Comment by ljeonjko** [comment]
Hi all!

I'm having trouble with running DINEOF.  Regardless of the size of the time dimension, I get the 

```
Diffusion of covariance matrix activated
 with parameters: alpha =            0.10    and number of iterations =   3

 Missing data:                     84134 out of       84134 (100.00%)

 Number of cross validation points                                881
 ********************************************************************

Note: The following floating-point exceptions are signalling: IEEE_INVALID_FLAG
STOP Not enough data
```
even though there is available data. 

Please advise.

Thanks!

P.S.
The .nc and .init file are in the .zip file.

[DINEOF_test.zip](https://github.com/aida-alvera/DINEOF/files/9189812/DINEOF_test.zip)

**Reply by ctroupin** [reply]
Hello,

Thanks for the report and for providing the data with it!

I've tried the data you've provided and got the same error (which is always a good 1st step).
Without checking yet in detail: the error message states that all the data are _missing_:
> Missing data:                     84134 out of       84134 (100.00%)

so in this case it is impossible to perform an analysis.

I will try to find the issue later, but maybe have a look at the mask.

**Edit:** it seems the mask is OK
![Screenshot from 2022-07-27 11-08-29](https://user-images.githubusercontent.com/11868914/181208974-cf26ed21-8d35-455e-8f7d-afb726c9586a.png)

**Reply by ctroupin** [reply]
OK I've got something that seems to work: the fill value you use is NaN. I changed it to -999., using this `nco` command:
```bash
ncatted -a _FillValue,,o,f,-999. FINAL_2015_027_075_050.nc
```

and now the analysis works:
```bash
 ********************************************************************
 Numerical data read
 You entered the values:
 number of EOF modes you want to compute          14
 maximal size for the Krylov subspace          19

 You asked not to normalise the input matrices

 You asked not to write the left and right EOFs used for the reconstruction
 (to set this option ON, change eof=0 in file dineof.init)

 ********************************************************************

 You entered filenames FINAL_1/FINAL_2015_027_075_050.nc#chlor_a
                       FINAL_1/FINAL_2015_027_075_050.nc#mask

 initfilename: config_TEST.init                                                                                                                                                                                        

 ********************************************************************
 Now some statistics about your data:

  Number of mask land points:                                    1382
        Dimension of file  1:                    64  x    64  x    31


                        Mean:                                  2.1158
          Standard deviation:                                  1.2913

 ***
 Matrix loaded ... Land points extracted...

 Size of the matrix used in DINEOF:                     2714 x     31
 ***
 reading time fileFINAL_1/FINAL_2015_027_075_050.nc#time                                                                                                                                                                                                                          
 time file read
 Diffusion of covariance matrix activated
 with parameters: alpha =            0.10    and number of iterations =   3

 Missing data:                     36029 out of       84134 ( 42.82%)

 Number of cross validation points                                881
 ********************************************************************

 Time (in seconds) for 1 EOF mode calculation in DINEOF         0.0024

 # EOF modes asked:  14        Convergence level required: 0.1E-02

 Starting with the EOF mode calculation...

 EOF mode    Expected Error    Iterations made   Convergence achieved
 ________    ______________    _______________   ____________________

     10              0.6913                350             0.2595E-02
     11              0.6637                350             0.5536E-02
     12              0.6716                350             0.5923E-02
     13              0.6564                350             0.5698E-02
     14              0.6631                350             0.7011E-02

  Minimum reached in cross-validation
  Number of optimal EOF modes:           13

  Make last reconstruction, including data put aside for cross-validation

     13              0.6564                350             0.5698E-02

 DINEOF finished!

number of eigenvalues retained for the reconstruction  13
         expected error calculated by cross-validation          0.6564
            total time (in seconds) in lanczos process          4.4093

 Now writing data...


 total variance of the initial matrix    1.6675117788026508     
 total variance of the reconstructed matrix    1.9303216157504632     

 ...done!
```

**Reply by ljeonjko** [reply]
Hi! 

Thanks so much for replying. I was thinking it had something to do with the values themselves, I'll give it a try again using your solution, I'll let you know how that goes. Thanks so much!

**Reply by ljeonjko** [reply]
It works! Thanks!

Just another question, is there a best way to determine  the

`nev 
neini 
ncv `

variables?

**Reply by ctroupin** [reply]
I would say:
- neini = 1
- nev: depends on the number of time steps you have in your data. One cannot compute more temporal EOFs than the number of time periods in the input file. For applications I've done, I think nev < 30 in general
- ncv > nev+5 so that's not hard to

**Reply by ljeonjko** [reply]
Thanks! Will implement this in work. Cheers.

**Comment by zhr9709** [comment]
Hello, when i try to run the similar dineof program based on matlab and git, i found the problem that Error with _saupd, info = -3
Check documentation in _saupd, i know how to solve it, but i found that there is no "time" in my monthly data, so the dimension of my file is 312 x 264 x1, is that means the error will absolutely happen？

thanks
Best regards

**Reply by ctroupin** [reply]
Hello,
if I understand well, your data matrix is 312 x 264 x1, so there is only one time instance in the dataset, hence DINEOF cannot perform the reconstruction. Is it correct?

**Reply by zhr9709** [reply]
Hello! Thank you very much for the reply. 
I think you understand correctly, because I have only one time instance in my data, so the data cannot be reconstructed using DINEOF. 
But I am using monthly or seasonal composite data and I think it should have about 30 or 90 time instances in one data, may I ask if there is any solution for this problem or I must change the raw data.
Attached is the data I use now.
[YOC.zip](https://github.com/aida-alvera/DINEOF/files/9796079/YOC.zip)


Yours sincerely
Zhang

**Comment by zhr9709** [comment]
Hello, everyone. I found a problem that my DINEOF based on matlab could not find the missing data out of the whole data, I have checked the mask and it was right.  I also found a problem about Error with _seupd, info =-2.May I ask if there is some solution for this problem.

![image](https://user-images.githubusercontent.com/115602704/199865661-d9ad83b7-44a9-4c91-bc07-11b3133ab119.png)
![image](https://user-images.githubusercontent.com/115602704/199866324-feb66c12-d02f-4f44-a690-2400de507f9a.png)


[DINEOFTSET.zip](https://github.com/aida-alvera/DINEOF/files/9934219/DINEOFTSET.zip)
Best regards

**Reply by ctroupin** [reply]
Hello,
I was wondering if there could be "NaN" values in your matrices.
Can you check that? I will download and test the data as well.

**Reply by ctroupin** [reply]
OK so the data files you sent are MODIS Level-3 chlorophyll concentration.
From that you can build the mask. Can you send the mask or an image showing it?

**Reply by zhr9709** [reply]
Hello, thank you for the reply. I have solve the problem and it was caused by the wrong mask I made. Now, I also have a little question that if my data is 2250X2219X228, can DINEOF process this large data? 

Best regards

**Reply by zhr9709** [reply]
> OK so the data files you sent are MODIS Level-3 chlorophyll concentration. From that you can build the mask. Can you send the mask or an image showing it?

Hello, thank you for the reply. I have solved the problem and it was caused by the wrong mask I made. Now, I also have a little question that if my data is 2250X2219X228, can DINEOF process this large data?

Best regards

**Reply by ctroupin** [reply]
2250X2219X228 is quite a large dataset, personally I don't think I have ever worked with this type of problem size, maybe @aida-alvera did. Anyway you can always try and it's too big it will stop pretty quickly.

**Reply by zhr9709** [reply]
> 2250X2219X228 is quite a large dataset, personally I don't think I have ever worked with this type of problem size, maybe @aida-alvera did. Anyway you can always try and it's too big it will stop pretty quickly.

Hello, thank you for the reply and advice. I will try it. Again thank you for providing us a great method of data interpolating.

Best regards
Zhang

**Comment by danandwen** [comment]
Hello, I want to process data in netcdf format, but I don't know how to run this code in linux system. I want to ask, how should I run it?

**Reply by ctroupin** [reply]
Hello,
on Linux you need to 
1) compile the code and 
2) run it using the command `dineof dineof.ini`

The compilation is explained in the README: https://github.com/aida-alvera/DINEOF,
it is performed using `make`. 
Then for an example you can try to run the one located in the [SmallExample](https://github.com/aida-alvera/DINEOF/tree/master/SmallExample) directory.

**Reply by danandwen** [reply]
Hello, my computer can run smallExample, but I don't know how to use this code in my data now, (my data is modis MCD43A4, nc format) I need to write in my data folder a  “xxxx.Init” document, and then run this “xxxx.init” document in linux?

**Reply by danandwen** [reply]
Excuse me, to use the code of dineof, I must have the sea and land mask of my data, right?and， I also want to ask a question. Excuse me, to use the code of dineof, I must have the sea and land mask of my data, right? If there is no sea and land mask, will my ocean also be interpolated to the data.

**Reply by ctroupin** [reply]
yes the idea is that:
- you have the data in the netCDF file, similar to the file in SmallExample
- you take the example file `dineof.init` and adapt it to your application: file name, variable name, etc
- for the land-sea mask: I think you need to at least create a matrix full of 1's that will be your _mask_, i.e., all the pixels are considered as _sea points_. If points are on land it's not very good because you never have measurements there, so the reconstruction won't make sense.

**Reply by danandwen** [reply]
![8NGP_37((MA SRWN3K M0HL](https://user-images.githubusercontent.com/103423671/203060388-662340a5-f60d-40ff-ab47-d56092c0603a.png)

I would like to ask, does this #SST need to be replaced with the variable name of my nc data, that is, my own data? If I have many variables in an nc file that need to be manipulated, how should I write it? Is it written like the following? data = ["test1.nc#band1#band2#band3"]

**Reply by ctroupin** [reply]
Maybe first you can test on one variable at a time. 
So yes, you need to replace 'SST' by the name of the variable in your netCDF file.

**Reply by danandwen** [reply]
Thank you so much, after these days of hard work, I can finally achieve this! ! Thank you very much, I would like to ask, what should I do if there is no longitude, latitude and time in the generated data, the generated data only has one variable name. How can I add code in init to generate the same latitude and longitude and time as the original data?
[red_filled.zip](https://github.com/aida-alvera/DINEOF/files/10090294/red_filled.zip)
here is my data

**Reply by danandwen** [reply]
[1.zip](https://github.com/aida-alvera/DINEOF/files/10090302/1.zip)
This compressed package is my original data。

**Reply by ctroupin** [reply]
good to know it works now!
Maybe the easiest solution, to get the coordinates etc, could be to write the reconstructed field in the same netCDF file as the input data. 

If we take the "SmallExample": in `dineof.init` you define:
- the input data: `data = ['alboranSST_small.nc#SST']`
- the output data: it is possible to write the reconstructed field in the same file (with another variable name), for example:
`results = ['alboranSST_small.nc#SST_filled']`

does it help?

**Comment by hidayatmn378** [comment]
Hi! I am Nur. I am trying to interpolate chlorophyll-a data from MODIS using DINEOF (Julia script), but somehow when I executed the

```
 #edit the path to the dineof executable!

dineof_exec = expanduser("~/DINEOF/dineof")
run(`$dineof_exec dineof_chla_modis_new.init`); 
```
This problem appears
```
 ******** Attention! ************************************************
 ********************************************************************
                                               Alpha =          0.1000
                               Alpha should not exceed          0.0000
 
 Alpha value is too large: the filter is unstable, DINEOF will stop
```
Can anyone help me?

I attach my code and the data that I used. Thanks. 

Nur 
[data.zip](https://github.com/aida-alvera/DINEOF/files/10472661/data.zip)

**Reply by aida-alvera** [reply]
Hi Nur,

if you have a dt of 0 in your data set (same day/hour) you will get this error, as the minimum alpha acceptable value depends on the minimum time step in your data.
If you have indeed 2 time frames at the exact very moment I would suggest combining them into a single image (compositing). If you prefer to keep those as separate images, then you cannot use the temporal filter (which is nevertheless suggested), so put alpha = 0 in the init file.
Hope this helps,
Aida

**Reply by hidayatmn378** [reply]
Thank you so much for your helpful explanation. Now it works.

**Comment by superakfjgjd** [comment]
When I run dineof with 160*200*8399 data, the program reports an error.The error message is：
Fortran runtime error: Attempt to allocate negative amount of memory. Possible integer overflow
My operating system is 64-bit and I use an already generated exe file.

**Reply by ctroupin** [reply]
Hell, can you also indicate the size of the matrix (number of times, latitudes and longitudes), just to have an idea.

Did you managed to run the Examples provided with the code?

**Reply by superakfjgjd** [reply]
I'm sorry, it's missing symbols.The time number is 8399, the latitude is 200, and the longitude is 160.
I suspect it exceeds 2,147,483,648(2^31).

**Reply by ctroupin** [reply]
thanks. To be sure this is the problem, maybe try with a lowest number of time instances, so you see when it breaks.

**Reply by superakfjgjd** [reply]
When I reduced the size, it worked.
I think the exe file is 32-bit, if so, is there a 64-bit one, I am a windows system.

**Reply by ctroupin** [reply]
Ok, I'm not sure we have the exe for 64-bit, I guess you have already tried to compile the source code?
Otherwise maybe @jmbeckers has it.

I'm working with Ubuntu but could give a try with Windows if needed.

**Reply by jmbeckers** [reply]
Sorry, I do not have an dineof.exe file for windows anymore. There was a problem in the past to make them with netCDF support
https://github.com/aida-alvera/DINEOF/issues/4

I do not know if that was solved in the meanwhile, but the best thing is probably to either try a compilation for windows from scratch or use the Ubuntu subsystem in Windows ?

**Reply by ctroupin** [reply]
Thanks!
I will try later on today and let you know.

Maybe we could use a _Github action_ to run the compilation for different OS
https://github.com/marketplace/actions/setup-fortran

**Reply by superakfjgjd** [reply]
Yes, I tried to compile the source code, but something went wrong in the middle.
Thank you for your reply and help.

**Reply by ctroupin** [reply]
I've tried to compile for Windows under Linux (with mingw-w64) unfortunately it doesn't go smoothly.

**Reply by jmbeckers** [reply]
Even without netCDF support ?

**Reply by superakfjgjd** [reply]
yes.

**Comment by arnabp893** [comment]
Hello all,
My name is ARNAB PAUL and I am a research assistant in Louisiana State University. 
I am working with MODIS L2 data and I am new in this community. I want to apply DINEOF to my Netcdf files but somehow I am having trouble working with it. I am using macOS terminal to download the DINEOF files from Github. any help will appreciated. Thank you.
I have attached a screenshot of output with this comment.
<img width="567" alt="Screenshot 2023-08-23 at 1 35 41 PM" src="https://github.com/aida-alvera/DINEOF/assets/66524822/2a8ad7d2-76a1-46d4-bafe-5bd62e32e13e">

**Reply by ctroupin** [reply]
Hello,

it looks like there is something missing in order to have netCDF working.

Can you indicate how the netCDF library was installed?
(I'm not too familiar with macOS).

Maybe something like 
```
sudo port install netcdf
```

**Reply by arnabp893** [reply]
sudo port install netcdf is not working. 
**brew install gcc
brew install make
brew install arpack
brew install netcdf**
This how I installed netcdf
I use Mac hence I am using Mac terminal.
if there is any other way to do I can opt for that like using gfortran in the system.

**Reply by ctroupin** [reply]
thanks.
Have you tried
```
brew install netcdf-fortran
```

**Reply by arnabp893** [reply]
Yes I did but the error remained the same. Something to do with netcdf.inc
after installation I am writing this in terminal
**git clone https://github.com/Aida-Alvera/DINEOF
cd DINEOF/
cp config.mk.template config.mk %copy the configuration file
make % create executable files**
Is there anything wrong with this snippet?

**Reply by jmbeckers** [reply]
I think you need to adapt the config.mk for your Mac System (see https://github.com/aida-alvera/DINEOF/tree/master/Compilers for other systems), but you might ask an expert on Mac compilers what to do in the config file.

**Reply by arnabp893** [reply]
I am not very good at those stuffs. How can I adapt the config.mk?
also I have a question that if I use Windows on my Mac with Boot Camp. Will it work?

.

**Reply by ctroupin** [reply]
It may work but usually running an OS within another OS can be an additional source of potential problems, in my experience.
The solution suggested by @jmbeckers is not that difficult, the idea is to edit a single file according to your installation.

We will try to see if other users managed to use DINEOF with MacOS, unfortunately nobody in the team has a Mac now.

**Comment by ljeonjko** [comment]
Hi all! I'm having trouble with DINEOF. I'm reconstructing data from two areas. 

This is the original data:
![image](https://github.com/aida-alvera/DINEOF/assets/104908008/6ccae1cd-e4de-4101-b0e6-b07108825ffe)

This is the reconstructed data:
![image](https://github.com/aida-alvera/DINEOF/assets/104908008/7b70b876-b423-452e-89c6-4a49534eb75b)

If you look closely, there's a discontinuity around x = 100. This discontinuity is persistent in time. Is this effect something to be expected with DINEOF? Thanks!

**Reply by ctroupin** [reply]
no, this effect is not expected with DINEOF.

Could you show a picture of the land-sea mask used for this application?

**Reply by ljeonjko** [reply]
Hi, thanks for the response!

Here is the mask:
![image](https://github.com/aida-alvera/DINEOF/assets/104908008/1ff983ac-8cc6-4a6c-9e58-00a9d82e0857)

Looking closely, I notice there are some pixels left unattended at around y = 125 (to clarify, x is vertical, y is horizontal).
Could these pixels be the cause of this discontinuity? 

Here's a zoomed-in image of the discontinuity - different variable, but the same area is problematical: 
![image](https://github.com/aida-alvera/DINEOF/assets/104908008/484ec8b7-1fd1-4eff-9bbf-4675badb01db)

Could increasing the nev (currently set at 14 for around 250 timesteps), solve this issue?

**Reply by ctroupin** [reply]
ok now I see better what you meant, at the beginning it was written `x=100` but it was around `y=100`, isn't it?

'nev' is the maximum number of modes to computed, you could increase it if the method doesn't converge, i.e. doesn't find an optimal number of EOFs. Otherwise I don't think it would make a difference.

How many modes are kept for the final reconstruction?
(you can find it in the file `neofretained.val`)

**Reply by ljeonjko** [reply]
For 

```
nev = 14
neini = 1
ncv = 19
```

there are `14` modes as per the file. Does this hint towards incomplete training?

**Reply by ctroupin** [reply]
Yes I think that if the number of EOFs used for the final reconstruction is equal to `nev`, then it means there wasn't a minimum reached. However, even if you increase `nev`, I'm not sure it will explain why you get the discontinuity.

But you can try a larger value to see, I guess it's not too long to complete. 

What method do you use for the selection of validation points? Random or extracted from existing clouds?

**Reply by ljeonjko** [reply]
Thanks, I'll try again with a higher  `nev` (30ish should be enough?) and see what happens.


I'm not sure, here is the `config.init` I'm using:

```
data = ["chlor_a_1998_MASKED.nc#chlor_a_masked"]
mask = ["chlor_a_1998_MASKED.nc#mask"]
time = "chlor_a_1998_MASKED.nc#time"
alpha = 0.1
numit = 3
nev = 14
neini = 1
ncv = 19
tol = 1.0e-8
nitemax = 350
toliter = 1.0e-3
rec = 0
norm = 0
eof = 0
Output = "."
results = [\chlor_a_1998_FILLED.nc#chl_filled"]
seed = 243435
EOF.U = ["./eof0.nc#Usst"]
EOF.V = "./eof0.nc#V"
EOF.Sigma = "./eof0.nc#Sigma"
```

What I didn't mention is that I'm using 25 years of data, but running DINEOF on year-long slices... but the issue occurs every year. Could the two areas (north and south of the discontinuity) be too different from each other, hence the discontinuity?

**Reply by ctroupin** [reply]
in my experience, DINEOF normaly generates such discontinuities, except is there is already a discontinuity in the original dataset. Have you checked that?

Also, if you have 25 years of data, if your computer allows it, don't hesitate to run DINEOF on the whole time period, you may get better reconstructions.

**Reply by ljeonjko** [reply]
thanks for the advice. 
as far as I can tell, there's not such discontinuity in the original data.
I've attempted 25 years at once, but it took way to long. I'll increase the `nev` and create chunks of 3,4 years. I'll post results in a day or two. 
cheers!

**Comment by aida-alvera** [comment]
Hello all,

I would advise to take a close look at the initial data. That structure
looks as it is originally in the data. A simple average might already show
it.

Two things that surprise me: the Gargano peninsula appears to be also
strange, like too straight in the east-west direction. Also, you say this
is chlorophyll? The values are very low and very patchy... what is the
source of the data?

Cheers,
Aida

**Reply by ljeonjko** [reply]
Hi, thanks for the response! 

I've examined the data, as far as I can tell, there is no such discontinuity in the data. I'll try again with the averages and see what comes out, but I'm fairly certain the inital data is clear. 
I have noticed the same thing with the Gargano peninsula, but I suppose it has something to do with the inital masking preparation for DINEOF. 
The data is remote sensing reflectance at 510, but the interesting thing is the discontinuity appears across all 6 bands (and the derived chlorophyll-a data as well). Data is OC-CCI V.6.0. 

I'm currently running DINEOF again against batches of 5-years-worth of data with increased number of EOFs, I'll post the results in a day or two when it's done. 

Cheers,

Leon

**Reply by ljeonjko** [reply]
False alarm, turns out there is a discontinuity in the data, I've noticed it when I checked the mean. Thanks again for the suggestion!

**Comment by superakfjgjd** [comment]
Hi,
When I use the MATLAB version of DINEOF, the website is https://svn.oss.deltares.nl/repos/openearthtools/trunk/matlab/applications/ +Dineof/, prompt function or variable 'mergestructures' not recognized

May I ask if this function was written by oneself? Can you provide it?

**Reply by ctroupin** [reply]
Sorry for the late reply, I hadn't seen your post.
If I'm not wrong those MATLAB files are not written by the GHER. 

There is an [example netCDF file](https://github.com/aida-alvera/DINEOF/blob/master/SmallExample/alboranSST_small.nc) that shows you how the input file should look like.

**Comment by arnabp893** [comment]
Hi I am Arnab and I have recently installed DINEOF. I am looking for directions of how to start processing my data. I have gappy satellite data from the MODIS.
Any help will be appreciated, thank you

**Reply by ctroupin** [reply]
I think the easiest way it to start from the example, in the folder `SmallExample` there is a file `alboranSST_small.nc` that you can use as an example. 

Then you need to write your MODIS data, for the region of interested, in a single netCDF file similar to  `alboranSST_small.nc`. In the example the file has only 10 time instances but for an application it is better to have more.

**Reply by arnabp893** [reply]
I have yearly data. How can use DINEOF to multiple images? Do you suggest MatLab? Or terminal?

**Reply by ctroupin** [reply]
With yearly data you won't have many time periods I think.

Better daily or maybe monthly.

> How can use DINEOF to multiple images? Do you suggest MatLab? Or terminal?

It depends what you are familiar with. Python and Julia also works. We cannot really suggest something.

**Reply by arnabp893** [reply]
I ran the alboranSST._small.nc as 
arnab@Arnab:/mnt/c/Users/arnab/DINEOF/SmallExample$ dineof dineof.init
and it worked. 
now how to use this for multiple .nc files? do I have to create a text file?

**Reply by ctroupin** [reply]
no, you need to write all the content of the netCDF files into a unique file.

**Comment by liuminqi-wyy** [comment]
Hello, I am an undergraduate student from China. See Scripts file there is a visualization program after interpolation results, I would like to ask what command should be issued on the linux terminal to achieve visualization? Or do you have another way to achieve this? I am new to linux, if you can answer my questions I would be grateful!

**Reply by ctroupin** [reply]
As the output of `DINEOF` are written as netCDF file(s), you can visualise them with any software tool that can read netCDF.
An easy-to-use tool is `ncview`, which you can install from https://cirrus.ucsd.edu/ncview/ and then use with commands like
```bash
ncview myncfile.nc
```

**Comment by JMPavs** [comment]
Help dineof.init for multivariate setup 
I'm trying to set up my dineof.init for multivariate but the output is getting read as an input file
<img width="1030" height="756" alt="image" src="https://github.com/user-attachments/assets/600ebab2-2063-4e49-b622-1ce867f0ffda" />
<img width="656" height="413" alt="image" src="https://github.com/user-attachments/assets/dac49979-dbd5-46fd-b579-e018de10bcfc" />

**Reply by JMPavs** [reply]
i have already resolved the issue thank you
